File: src/Web/JavaScriptJS/patterns/memento.js
const assert = require("node:assert/strict");
/** Captures and restores editor state without exposing restoration internals. */
function run() {
  const editor = {
    text: "",
    snapshot() { return Object.freeze({ text: this.text }); },
    restore(memento) { this.text = memento.text; },
  };
  editor.text = "draft";
  const saved = editor.snapshot();
  editor.text = "broken";
  editor.restore(saved);
  assert.equal(editor.text, "draft");
  return "memento";
}
module.exports = { run };
if (require.main === module) run();

File: src/Scripting/Ruby/patterns/memento.rb
# frozen_string_literal: true

module RubyPatterns
  module Memento
    module_function

    def run
      state = { text: 'draft' }
      snapshot = state.dup.freeze
      state[:text] = 'edited'
      state.replace(snapshot)
      raise 'Memento failed' unless state[:text] == 'draft'
      true
    end
  end
end

RubyPatterns::Memento.run if $PROGRAM_NAME == __FILE__
